rust/src/dumpfile.rs

```diff
@@ -0,0 +1,391 @@
+use std::fmt::Write as WriteFmt;
+use std::io::Write;
+use std::os::unix::ffi::{OsStrExt, OsStringExt};
+use std::path::{Path, PathBuf};
+use std::str::FromStr;
+
+use anyhow::Context;
+use anyhow::{anyhow, Result};
+
+#[derive(Debug, PartialEq, Eq)]
+/// An extended attribute entry
+pub struct Xattr<'k> {
+    /// key
+    pub key: Cow<'k, OsStr>,
+    /// value
+    pub value: Cow<'k, [u8]>,
+}
+/// A full set of extended attributes
+pub type Xattrs<'k> = Vec<Xattr<'k>>;
+
+/// Modification time
+#[derive(Debug, PartialEq, Eq)]
+pub struct Mtime {
+    /// Seconds
+    pub sec: u64,
+    /// Nanoseconds
+    pub nsec: u64,
+}
+
+/// A composefs dumpfile entry
+#[derive(Debug, PartialEq, Eq)]
+pub struct Entry<'p> {
+    /// The filename
+    pub path: Cow<'p, Path>,
+    /// uid
+    pub uid: u32,
+    /// gid
+    pub gid: u32,
+    /// mode (includes file type)
+    pub mode: u32,
+    /// Modification time
+    pub mtime: Mtime,
+    /// The specific file/directory data
+    pub item: Item<'p>,
+    /// Extended attributes
+    pub xattrs: Xattrs<'p>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+/// A serializable composefs entry.
+///
+/// The `Display` implementation for this type is defined to serialize
+/// into a format consumable by `mkcomposefs --from-file`.
+pub enum Item<'p> {
+    /// A regular file
+    Regular {
+        /// Size of the file
+        size: u64,
+        /// Number of links
+        nlink: u32,
+        /// Inline content
+        inline_content: Option<Cow<'p, [u8]>>,
+        /// The fsverity digest
+        fsverity_digest: Option<String>,
+    },
+    /// A character or block device node
+    Device {
+        /// Number of links
+        nlink: u32,
+        /// The device number
+        rdev: u32,
+    },
+    /// A symbolic link
+    Symlink {
+        /// Number of links
+        nlink: u32,
+        /// Symlink target
+        target: Cow<'p, Path>,
+    },
+    /// A hardlink entry
+    Hardlink {
+        /// The hardlink target
+        target: Cow<'p, Path>,
+    },
+    /// FIFO
+    Fifo {
+        /// Number of links
+        nlink: u32,
+    },
+    /// A directory
+    Directory {
+        /// Size of a directory is not necessarily meaningful
+        size: u64,
+        /// Number of links
+        nlink: u32,
+    },
+}
+
+/// Unescape a byte array according to the composefs dump file escaping format.
+fn unescape(s: &str) -> Result<Cow<[u8]>> {
+    // If there are no escapes, just return the input unchanged
+    if !s.contains('\\') {
+        return Ok(Cow::Borrowed(s.as_bytes()));
+    }
+    let mut it = s.chars();
+    let mut r = Vec::new();
+    while let Some(c) = it.next() {
+        if c != '\\' {
+            write!(r, "{c}").unwrap();
+            continue;
+        }
+        let c = it.next().ok_or_else(|| anyhow!("Unterminated escape"))?;
+        let c = match c {
+            '\\' => b'\\',
+            'n' => b'\n',
+            'r' => b'\r',
+            't' => b'\t',
```

Comment:
We also support `\=` to escape =

Reply:
Oh, or at least the docs say, but we don't seem too actually do it.

Reply:
We should probably fix the docs here.

Reply:
Done in https://github.com/containers/composefs/pull/287